src/borrow_tracker/tree_borrows/wildcard.rs

```diff
@@ -0,0 +1,409 @@
+        let has_foreign = self.max_foreign_access >= WildcardAccessLevel::Read;
+        let has_child = self.child_reads > 0 || exposed_as >= WildcardAccessLevel::Read;
+        use WildcardAccessRelatedness as E;
+        match (has_foreign, has_child) {
+            (true, true) => Some(E::EitherAccess),
+            (true, false) => Some(E::ForeignAccess),
+            (false, true) => Some(E::ChildAccess),
+            (false, false) => None,
+        }
+    }
+    /// from where relative to this pointer a write access could happen
+    pub fn write_access_relatedness(
+        &self,
+        exposed_as: WildcardAccessLevel,
+    ) -> Option<WildcardAccessRelatedness> {
+        let has_foreign = self.max_foreign_access == WildcardAccessLevel::Write;
+        let has_child = self.child_writes > 0 || exposed_as == WildcardAccessLevel::Write;
+        use WildcardAccessRelatedness as E;
+        match (has_foreign, has_child) {
+            (true, true) => Some(E::EitherAccess),
+            (true, false) => Some(E::ForeignAccess),
+            (false, true) => Some(E::ChildAccess),
+            (false, false) => None,
+        }
+    }
+    /// gets the access tracking information for a new child node.
+    /// doesnt have any child reads/writes, but calculates max_foreign_access from parent
+    pub fn get_new_child(&self, exposed_as: WildcardAccessLevel) -> Self {
+        Self {
+            max_foreign_access: max(self.max_foreign_access, self.max_child_access(exposed_as)),
+            child_reads: 0,
+            child_writes: 0,
+        }
+    }
+    /// update the tracking information of a tree, to reflect the access level change from `old_access_type` to `access_type` of an exposed pointer.
+    /// propagates the wilcard access information over the tree
+    /// this needs to be called every time the access level of an exposed reference changes, to keep the state in sync
+    pub fn update_exposure(
+        id: UniIndex,
+        old_access_type: WildcardAccessLevel,
+        access_type: WildcardAccessLevel,
+        nodes: &UniValMap<Node>,
+        perms: &UniValMap<LocationState>,
+        wildcard_accesses: &mut UniValMap<WildcardAccessTracking>,
+    ) {
+        /// pushes children onto the stack, if their `max_foreign_access` field needs to be updated
+        ///
+        /// the `max_foreign_access` fields is set based on the max of the parents `max_foreign_access`,
+        /// `exposed_as` and its siblings `max_child_access`.
+        ///
+        /// this function calculates the siblings `max_child_access`, both of the other fields need to be passed as arguments
+        ///
+        /// * `other_factors`:  we only ever change one of these values. The max value of the other fields we dont change should be passed through the `other_factors` parameter
+        /// * `old_access_type`,`access_type`: we change the parameter not covered by `other_factors` from `old_access_type`
```

Comment:
"we change" seems to mean "the caller changed/changes" or so? It sounds like this function is doing the changing.